### PR TITLE
fix: Avoid side effect

### LINKED ISSUE #531: [bug] Side effect in Trial controller spawnMetricsCollector
/kind bug

Run trial controller and create a trial, the metrics collector will be created multiple times. We do not check if the resource is created in the cluster. Thus it will return an error and run the endless reconcile loop

/assign

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Kubeflow version:
- Minikube version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by gaocegege**
Signed-off-by: Ce Gao <gaoce@caicloud.io>

**What this PR does / why we need it**:

Fix #531 

We will meet an error during the creation and retry it again and again. I strongly suggest using state compare pattern instead of current state machine based implementation. But it is another topic.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/532)

### CODE REVIEW

File: pkg/controller/v1alpha2/trial/trial_controller.go [hunk 1]
```diff
@@ -224,24 +234,20 @@ func (r *ReconcileTrial) reconcileJob(instance *trialsv1alpha2.Trial, desiredJob
...
 		if errors.IsNotFound(err) {
-			logger.Info("Creating Job", "kind", kind)
+			logger.Info("Creating Job", "kind", kind,
+				"name", desiredJob.GetName())
 			err = r.Create(context.TODO(), desiredJob)
 			if err != nil {
 				logger.Error(err, "Create job error")
 				return nil, err
 			}
+			deployedJob = desiredJob
```
Comment (johnugeorge) [inline pkg/controller/v1alpha2/trial/trial_controller.go]: Is this needed?  deployed job is returned if job is already in running state. In reconcile function(https://github.com/kubeflow/katib/pull/532/files#diff-63bd3a953ae0c850443cefebf8365018R21), update status if deployed job is not null

**Comment by johnugeorge** [comment]
Thanks. This is a required change. I was about to give the  comment in https://github.com/kubeflow/katib/pull/496 before it got merged :)

File: pkg/controller/v1alpha2/trial/trial_controller.go (hunk 1 above)
Comment (gaocegege) [reply pkg/controller/v1alpha2/trial/trial_controller.go]: If we do not set it, deployedJob is an empty job without status. And we have to update it in the next reconcile loop. I think we can do it in this loop.

**Comment by gaocegege** [comment]
@johnugeorge Yeah my mistake. I did not find this potential problem. And when I am running unit test in #528 I find the logic has bugs.

File: pkg/controller/v1alpha2/trial/trial_controller.go (hunk 1 above)
Comment (gaocegege) [reply pkg/controller/v1alpha2/trial/trial_controller.go]: WDYT, actually we keep an eventually consistent model, thus it is not a critical change.

Comment (gaocegege) [reply pkg/controller/v1alpha2/trial/trial_controller.go]: Let's remove it since I think it is not a necessary change.

**Comment by gaocegege** [comment]
@johnugeorge PTAL

File: pkg/controller/v1alpha2/trial/trial_controller.go (hunk 1 above)
Comment (johnugeorge) [reply pkg/controller/v1alpha2/trial/trial_controller.go]: OK. agree

**Comment by gaocegege** [comment]
I am not why why bot does not merge the PR automatically.

**Comment by johnugeorge** [comment]
Tests failed @gaocegege

**Comment by gaocegege** [comment]
So the CI will be run again after we approve one PR.

**Pull request merged**